processor/processor.go

```diff
@@ -20,44 +20,66 @@ package processor
 import (
 	"github.com/santhosh-tekuri/jsonschema"
 
-	"github.com/elastic/apm-server/model"
+	"github.com/elastic/apm-server/decoder"
+	"github.com/elastic/apm-server/model/metadata"
+	"github.com/elastic/apm-server/transform"
+	"github.com/elastic/apm-server/utility"
 	"github.com/elastic/apm-server/validation"
 	"github.com/elastic/beats/libbeat/monitoring"
 )
 
 type Processor interface {
 	Validate(map[string]interface{}) error
-	Decode(map[string]interface{}) (model.Payload, error)
+	Decode(map[string]interface{}) (*metadata.Metadata, []transform.Transformable, error)
 	Name() string
 }
 
-type PayloadDecoder func(map[string]interface{}) (model.Payload, error)
-
-type PayloadProcessor struct {
-	ProcessorName string
-	DecodePayload PayloadDecoder
+type EventsProcessor struct {
+	SingularName  string
+	EventDecoder  decoder.EventDecoder
 	PayloadSchema *jsonschema.Schema
 	DecodingCount *monitoring.Int
 	DecodingError *monitoring.Int
 	ValidateCount *monitoring.Int
 	ValidateError *monitoring.Int
 }
 
-func (p *PayloadProcessor) Name() string {
-	return p.ProcessorName
+func (p *EventsProcessor) Name() string {
+	return p.SingularName
+}
+
+func (p *EventsProcessor) decodePayload(raw map[string]interface{}) ([]transform.Transformable, error) {
+	var err error
+	decoder := utility.ManualDecoder{}
+
+	pluralName := p.SingularName + "s"
+	rawObjects := decoder.InterfaceArr(raw, pluralName)
```

Comment:
I'd have handled it the other way around, as the `singularized name` is only used in test code, but the pluralized isn't. But minor thing, I'm ok with this solution.